qualtran/bloqs/rotations/phase_gradient.py

```diff
@@ -293,36 +293,49 @@ def with_registers(self, *new_registers: Union[int, Sequence[int]]):
         raise NotImplementedError("not needed.")
 
     @cached_method
-    def scaled_val(self, x: int) -> int:
-        """Computes `phase_grad + x` using fixed point arithmetic."""
+    def scaled_val_int(self, x: int) -> int:
+        """Computes `x >> right_shift` using fixed point arithmetic."""
         x_width = self.x_bitsize + self.right_shift
         x_fxp = _fxp(x / 2**x_width, x_width).like(_fxp(0, self.phase_bitsize)).astype(float)
         return int(x_fxp.astype(float) * 2**self.phase_bitsize)
 
     def apply(self, *args) -> Tuple[Union[int, np.integer, NDArray[np.integer]], ...]:
         if self.controlled_by is not None:
             ctrl, x, phase_grad = args
-            out = self.on_classical_vals(ctrl=ctrl, x=x, phase_grad=phase_grad)
-            return out['ctrl'], out['x'], out['phase_grad']
-
-        x, phase_grad = args
-        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
-        return out['x'], out['phase_grad']
-
-    def on_classical_vals(self, **kwargs) -> Dict[str, 'ClassicalValT']:
-        x, phase_grad = kwargs['x'], kwargs['phase_grad']
-        if self.controlled_by is not None:
-            ctrl = kwargs['ctrl']
             if ctrl == self.controlled_by:
-                phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (
+                phase_grad_out = (phase_grad + self.sign * self.scaled_val_int(x)) % (
                     2**self.phase_bitsize
                 )
             else:
                 phase_grad_out = phase_grad
-            return {'ctrl': ctrl, 'x': x, 'phase_grad': phase_grad_out}
 
-        phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (2**self.phase_bitsize)
-        return {'x': x, 'phase_grad': phase_grad_out}
+            return ctrl, x, phase_grad_out
+
+        x, phase_grad = args
+        phase_grad_out = (phase_grad + self.sign * self.scaled_val_int(x)) % (
+            2**self.phase_bitsize
+        )
+        return x, phase_grad_out
+
+    def on_classical_vals(self, **kwargs) -> Dict[str, Union['ClassicalValT', Fxp]]:
+        x: Fxp = kwargs['x']
+        phase_grad: Fxp = kwargs['phase_grad']
+
+        if self.controlled_by is None or self.controlled_by == kwargs['ctrl']:
+            # widen appropriately so that right shifting does not drop necessary bits
+            x = x.like(
+                QFxp(x.n_int + phase_grad.n_frac, phase_grad.n_frac, phase_grad.signed)._fxp_dtype
+            )
+            scaled_x = x >> self.right_shift
```

Comment:
nit: Since `self.x_dtype` is `QFxp(self.x_bitsize, self.x_bitsize, signed=False)`, `x.n_int` is always 0 and so we can simply use `x.like(_fxp(self.phase_bitsize))` to do the resizing before shifting. 
```suggestion
            scaled_x = x.like(_fxp(self.phase_bitsize)) >> self.right_shift
```

Reply:
I was planning to remove `_fxp` and only use `QFxp` methods to construct Fxps so that there's only one way to construct them (easier to maintain). 

I'll use `x.like(self.phase_dtype._fxp_dtype)` here instead.

Reply:
The `_fxp` helper here sets useful properties on the `Fxp` config object (like `shifting='trunc'`) which are important to guarantee correctness of the multiply via add logic. This is hard to enforce when constructing `Fxp` objects via `QFxp`; since this is relevant in the context of this arithmetic bloq; please use the `_fxp` helper here instead of `self.phase_dtype._fxp_dtype` (which would be wrong the way it's implemented right now; unless we change the default options on `QFxp._fxp_dtype`

Reply:
I actually fixed the `_fxp_dtype` to match the `_fxp` implementation, as we'd almost always want that.